src/main/java/org/wildfly/security/sasl/digest/DigestSaslClient.java

```diff
@@ -57,20 +58,21 @@
     private final String[] demandedCiphers;
     private final MessageDigest messageDigest;
 
-    DigestSaslClient(String mechanism, String protocol, String serverName, CallbackHandler callbackHandler, String authorizationId, boolean hasInitialResponse, Charset charset, String[] ciphers) throws SaslException {
+    DigestSaslClient(String mechanism, String protocol, String serverName, CallbackHandler callbackHandler, String authorizationId, boolean hasInitialResponse, Charset charset, String[] qops, String[] ciphers) throws SaslException {
         super(mechanism, protocol, serverName, callbackHandler, FORMAT.CLIENT, charset, ciphers);
 
         this.hasInitialResponse = hasInitialResponse;
         this.authorizationId = authorizationId;
-        this.demandedCiphers = (ciphers == null ? new String[] {} : ciphers);
+        this.clientQops = qops == null ? new String[] {DigestUtils.QOP_AUTH} : qops;
+        this.demandedCiphers = ciphers == null ? new String[] {} : ciphers;
         try {
             this.messageDigest = MessageDigest.getInstance(DigestUtils.messageDigestAlgorithm(mechanism));
         } catch (NoSuchAlgorithmException e) {
-            throw new SaslException("Expected message digest algorithm is not available", e);
+            throw new SaslException(getMechanismName() + ": Expected message digest algorithm is not available", e);
```

Comment:
We probably need to be looking to get the mechanism error messages converted to ElytronMessages, this way instead of concatenating the values like this a formatted message can be used.  Also an unavailable digest error message seems like something that could be used across mechanism implementations.